base/iobuffer.jl

```diff
@@ -2,19 +2,75 @@
+#   | offset   |    ptr        size   |     |
+#   1         mark                    |     lastindex(data)
+#                                     maxsize
+
+#            AFTER COMPACTION
+# Mark, ptr and size decreases by (mark - offset - 1)
+
+#   ...uuuuuXXXXXXXXXXXXX--------------.......
+#   | ||    |           |             |      |
+#   | offset|           size          |      lastindex(data)
+#   1  |    ptr                       maxsize
+#      mark ( == offset + 1)
+
+# * The underlying array is always 1-indexed
+# * Data in 1:offset is never touched. This allows an IOBuffer to use a subset of a larger array.
+# * Data in maxsize+1:lastindex(data) is also never touched.
+# * Data in offset+1:mark-1, if mark > 0 can be deleted, shifting the whole thing to the left
+#   to make room for more data, without replacing or resizing data
+
 mutable struct GenericIOBuffer{T<:AbstractVector{UInt8}} <: IO
-    data::T # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
-    reinit::Bool # if true, data needs to be re-allocated (after take!)
+    # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
+    data::T
+
+    # The user can take control of `data` out of this struct. When that happens, instead of eagerly allocating
+    # a new array, we set `.reinit` to true, and then allocate a new one when needed.
+    reinit::Bool
     readable::Bool
     writable::Bool
-    seekable::Bool # if not seekable, implementation is free to destroy (compact) past read data
-    append::Bool # add data at end instead of at pointer
-    size::Int # end pointer (and write pointer if append == true) + offset
-    maxsize::Int # fixed array size (typically pre-allocated)
-    ptr::Int # read (and maybe write) pointer + offset
-    offset::Int # offset of ptr and size from actual start of data and actual size
-    mark::Int # reset mark location for ptr (or <0 for no mark)
+
+    # If not seekable, implementation is free to destroy (compact) data in offset+1:mark-1
+    seekable::Bool
+
+    # If true, write new data to the index size+1 instead of the index ptr.
+    append::Bool
+
+    # Last index of `data` that has been written to. Data in size+1:end has not yet been used,
+    # and may contain arbitrary values.
+    # This value is always in offset : min(maxsize, lastindex(data))
+    size::Int
+
+    # Size can never be larger than this value. This is useful for two use cases:
+    # 1. To prevent `data` from becoming too large, using too much memory
+    # 2. To guarantee that data in maxsize+1:lastindex(data) is never touched
+    # Note that this can be larger than lastindex(data), since data may be replaced.
+    # This value is always in 0:typemax(Int)
+    maxsize::Int
+
+    # Data is read/written from/to ptr, except in situations where append is true, in which case
+    # data is still read from ptr, but written to size+1.
+    # This value is alwaus in offset+1 : size+1
```

Comment:
This is actually a bug in the current implementation. The ptr is not supposed to be restricted to be inside the file size, and any calls to write that happen should then automatically zero-pad the file from size:ptr (https://pubs.opengroup.org/onlinepubs/009695399/functions/lseek.html)

Reply:
Hmm how does that work? Would the user have to `seek` first to beyond `size`? That doesn't correspond to how seek normally works in Julia: Normally, you can't seek beyond the filesize. Should that be allowed? Then what happens if the user seeks beyond `lastindex(buffer.data)`?

Reply:
How is it supposed to work with `IOStream`? What file should the following script create?
```julia
f = open("/tmp/foo", "w+")
write(f, "abcd")
flush(f)
seek(f, 10)
write(f, 'a')
flush(f)
```